### PR TITLE
feat(cli): add providers in parallel

### LINKED ISSUE #2110: cdktf provider add should add all providers in parallel and not abort if one of them fails
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I tried running `cdktfl provider add aws google azurerm cloudflare kubernetes --force-local` and if failed with cloudflare since that provider was not in the hashicorp namespace. all of the providers were added one after another. I think we should try to add them in parallel and at least not stop once one failed, the kubernetes provider could have been added successfully already.

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Closes #2110

### CODE REVIEW

File: packages/cdktf-cli/src/bin/cmds/handlers.ts [hunk 1]
```diff
@@ -417,21 +420,29 @@ export async function providerAdd(argv: any) {
...
     if (argv.forceLocal) {
-      needsGet = true;
       await manager.addLocalProvider(constraint);
+      return true;
     } else {
       const { addedLocalProvider } = await manager.addProvider(constraint);
-      if (addedLocalProvider) {
-        needsGet = true;
-      }
+      return addedLocalProvider;
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/handlers.ts:431]: As discussed offline, we should confirm that we don't run into race conditions like e.g. running `npm install` twice at the same time.

File: packages/cdktf-cli/src/bin/cmds/handlers.ts [hunk 2]
```diff
@@ -440,21 +445,41 @@ export async function providerAdd(argv: {
     }
     return errorMessage;
   }, "");
+
   const needsGet = results.some(
-    (item) => item.status === "fulfilled" && item.value === true
+    (item) =>
+      item.status === "fulfilled" && item.value.addedLocalProvider === true
   );
+  const packagesToInstall = results.reduce((packages, item) => {
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/handlers.ts:453]: nit: I think changing the `reduce` to a chained `filter` and `flatMap` would make this more readable 😁
#no-reduce 😄

File: packages/cdktf-cli/src/bin/cmds/handlers.ts [hunk 3]
```diff
@@ -440,21 +445,41 @@ export async function providerAdd(argv: {
...
+    if (item.status === "fulfilled" && item.value.prebuiltProviderToAdd) {
+      return [...packages, item.value.prebuiltProviderToAdd];
+    }
+    return packages;
+  }, [] as PackageConstraint[]);
 
+  const sideEffects: Promise<unknown>[] = [];
```
Comment (ansgarm) [inline packages/cdktf-cli/src/bin/cmds/handlers.ts:460]: hehe, this is kinda cool 😁

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 4]
```diff
@@ -12,6 +12,11 @@ import fetch from "node-fetch";
 import * as z from "zod";
 import { logger } from "../logging";
 
+export type PackageConstraint = {
+  name: string;
+  version?: string;
+};
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/package-manager.ts:18]: love it!

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 5]
```diff
@@ -141,22 +144,25 @@ class NodePackageManager extends PackageManager {
...
 
     // Install exact version
     // Yarn: https://classic.yarnpkg.com/lang/en/docs/cli/add/#toc-yarn-add-exact-e
     // Npm: https://docs.npmjs.com/cli/v8/commands/npm-install#save-exact
     args.push("-E");
 
-    console.log(
-      `Installing package ${packageName} @ ${packageVersion} using ${command}.`
+    logger.debug(
+      `Installing package ${packages
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/package-manager.ts:158]:
```suggestion
      `Installing package(s) ${packages
```

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 6]
```diff
@@ -238,30 +244,35 @@ class PythonPackageManager extends PackageManager {
...
+          },
+          stdio: ["inherit", 1, 1],
+        }
+      );
 
       console.log("Package installed.");
     } else {
       console.log(
-        `Installing package ${packageName} @ ${packageVersion} using pip.`
+        `Installing package ${packages
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/package-manager.ts:273]:
```suggestion
        `Installing package(s) ${packages
```

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 7]
```diff
@@ -290,16 +293,21 @@ export class DependencyManager {
     return packageVersion;
   }
 
-  async addPrebuiltProvider(constraint: ProviderConstraint) {
+  async addPrebuiltProvider(
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:296]: We should rename this method, as it is not adding anything anymore, but rather returns a `PackageConstraint` that can be added later on.

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 8]
```diff
@@ -137,12 +137,13 @@ export class DependencyManager {
     );
   }
 
-  async addProvider(
-    constraint: ProviderConstraint
-  ): Promise<{ addedLocalProvider: boolean }> {
+  async addProvider(constraint: ProviderConstraint): Promise<{
+    addedLocalProvider: boolean;
+    prebuiltProviderToAdd?: PackageConstraint;
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:142]: maybe name this `prebuiltProviderToInstall` to denote that this is a dependency in the target language's depedency manager? "Add" seems a bit overused in this context.

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 9]
```diff
@@ -137,12 +137,13 @@ export class DependencyManager {
     );
   }
 
-  async addProvider(
-    constraint: ProviderConstraint
-  ): Promise<{ addedLocalProvider: boolean }> {
+  async addProvider(constraint: ProviderConstraint): Promise<{
```
Comment (ansgarm) [inline packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:140]: nit: The method is a bit leaky now, as it either has no side-effect (when the provider ends up being available as pre-built provider) or it does have a side-effect (when it's added locally to the `cdktf.json`).
I could see us changing the method name to try to reflect the non-side-effect nature or to maybe refactor this in a way that changes to the local-adding bit to have no side-effect, too, and also return "installation instructions" aka a constraint that can be added later on.

Comment (ansgarm) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:140]: While it is pretty rare – in theory (e.g. with a slow filesystem) reading the local providers, filtering them, adding another local provider, and writing them might be a different race condition ([source](https://github.com/hashicorp/terraform-cdk/blob/950c919b901157673f5afda47dc0e4b7ecd3f231/packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts#L366)). Maybe refactoring that bit in a similar way to the pre-built provider packages would get rid of that race condition while solving my nit about this inconsistency / about the methods name 🤔

**Comment by DanielMSchmidt** [comment]
Closing this since fixing all the small race conditions makes the code harder to read and is not worth the effort for the little performance improvement we get out of it.

**Pull request closed without merging**